### PR TITLE
[WIP] Webpack implementation

### PR DESCRIPTION
**Pull request by mxstbr**
Switching to webpack for bundling due to HMR.

**TODO**
- [x] Production bundle
- [x] Development bundle
  - [ ] Hot module replacement

---

There's problems with two modules that actually aren't meant to be used on the client: `joi` and `numbro`. We use them, but they `require` Node.js internal modules like `fs` which don't exist on the client.

I fixed it by spoofing these modules with empty files for webpack, but I'm not sure how your browserify process is working since those aren't actually bundleable. This might've broken email validation (if we're doing that) with `joi`, but I'm not yet 100% sure, would appreciate somebody trying that where we're using it.

_See foretagsplatsen/numbro#74 and hapijs/joi#154 for the solutions I did._

### CODE REVIEW

File: frontend/config/webpack.prod.babel.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
+var path = require('path');
+var webpack = require('webpack');
+
+module.exports = {
```
Comment (sedubois) [inline frontend/config/webpack.prod.babel.js]: note that you could have a single webpack config file as well, and just configure the JSON config object differently depending either on `NODE_ENV` or on `npm_lifecycle_event`. Like here https://github.com/belongapp/belong-website/blob/master/webpack.config.js

Comment (mxstbr) [inline frontend/config/webpack.prod.babel.js]: Yeah I know, this is a first implementation just to get a feel for it. :grin:

I'll probably do it how we did it for `react-boilerplate` with a `webpack.base.babel.js`: https://github.com/mxstbr/react-boilerplate/tree/master/internals/webpack (the `if/else` in a single file gets unmaintainable when config gets more complex)

File: frontend/src/containers/HomePage.js [hunk 2]
```diff
@@ -48,7 +48,7 @@ export class HomePage extends Component {
               <use xlinkHref='#svg-logotype' fill='#303233'/>
             </svg>
           </div>
-          <div className='subtitle'>organizing the Internet generation</div>
+          <div className='subtitle'>organizing the Internet</div>
```
Comment (sedubois) [inline frontend/src/containers/HomePage.js]: intended?

File: package.json [hunk 3]
```diff
@@ -108,7 +109,8 @@
     "svg-to-png": "^3.1.0",
     "uglify-js": "2.6.2",
     "universal-analytics": "0.3.11",
-    "webpack": "^2.1.0-beta.20"
+    "webpack": "^2.1.0-beta.20",
+    "webpack-dev-server": "^1.14.1"
```
Comment (sedubois) [inline package.json]: better to freeze the dependencies

File: frontend/src/containers/HomePage.js (hunk 2 above)
Comment (mxstbr) [inline frontend/src/containers/HomePage.js]: nope, accidental 😁 thanks!

File: package.json (hunk 3 above)
Comment (mxstbr) [inline package.json]: Good catch!

**Comment by mxstbr** [comment]
This is still WIP, dev mode doesn't work because of the server rendering. I might be able to work around that, but I think it's worth talking about disabling SSR in dev maybe? /cc @asood123

File: package.json [hunk 4]
```diff
@@ -39,11 +40,13 @@
     "babel": "5.8.38",
     "babel-core": "6.7.7",
     "babel-eslint": "6.1.2",
+    "babel-loader": "^6.2.4",
```
Comment (sedubois) [inline package.json]: freeze dep and same for others in this file

File: package.json [hunk 5]
```diff
@@ -103,7 +108,9 @@
     "serve-favicon": "2.3.0",
     "svg-to-png": "^3.1.0",
     "uglify-js": "2.6.2",
-    "universal-analytics": "0.3.11"
+    "universal-analytics": "0.3.11",
+    "webpack": "2.1.0-beta.20",
+    "webpack-dev-server": "1.14.1"
```
Comment (sedubois) [inline package.json]: move to dev dependencies?

File: package.json [hunk 6]
```diff
@@ -39,11 +40,13 @@
     "babel": "5.8.38",
     "babel-core": "6.7.7",
     "babel-eslint": "6.1.2",
+    "babel-loader": "^6.2.4",
     "babel-plugin-add-module-exports": "0.1.4",
     "babel-plugin-lodash": "2.3.0",
     "babel-plugin-transform-es2015-modules-commonjs": "6.7.7",
     "babel-preset-es2015": "6.6.0",
     "babel-preset-react": "6.5.0",
+    "babel-preset-react-hmre": "^1.1.1",
```
Comment (sedubois) [inline package.json]: move to dev dependencies?

File: frontend/src/lib/env.js [hunk 7]
```diff
@@ -2,12 +2,14 @@
  * Environment variables
  */
 
-if (typeof process.env.NODE_ENV === 'undefined') {
-  process.env.NODE_ENV = 'development';
+let NODE_ENV = process.env.NODE_ENV;
+
+if (NODE_ENV === 'undefined') {
+  NODE_ENV = 'development';
```
Comment (asood123) [inline frontend/src/lib/env.js]: Doesn't this mean that process.env.NODE_ENV won't be set to anything?

Comment (mxstbr) [inline frontend/src/lib/env.js]: You can't set `process.env.NODE_ENV`, it throws an error and the entire application grinds to halt…

**Comment by asood123** [comment]
What about using this for 'Joi'? https://github.com/jeffbski/joi-browser

**Pull request closed without merging**